### PR TITLE
VDB-1629 - Have CheckedHeadersRepository use the plugin schema specified

### PR DESCRIPTION
**Pull request by paytonrules**
As part of pulling Oasis into its own repository (and by extension removing it from MCD-transformers) we can't share the same "check_count" on the headers table anymore. Plugins provide their schema and must have a matching table.

This also removes the need for CLIENT_IPCPATH from the make test command, which was an accident.

### CODE REVIEW

File: integration_test/contract_watcher_transformer_test.go [hunk 1]
```diff
@@ -29,13 +29,14 @@ import (
...
 	"github.com/makerdao/vulcanizedb/pkg/datastore/postgres/repositories"
+	"github.com/makerdao/vulcanizedb/test_config"
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
 )
 
 var _ = Describe("contractWatcher transformer", func() {
 	var (
-		db               *postgres.DB
+		db               = test_config.NewTestDB(test_config.NewTestNode())
```
Comment (rmulhol) [inline integration_test/contract_watcher_transformer_test.go:38]: seems like we're overwriting this assignment in the `BeforeEach`?

File: pkg/datastore/postgres/repositories/checked_headers_repository.go [hunk 2]
```diff
@@ -17,62 +17,77 @@
...
+
 	"github.com/makerdao/vulcanizedb/pkg/core"
 	"github.com/makerdao/vulcanizedb/pkg/datastore/postgres"
 )
 
 const (
-	insertCheckedHeaderQuery = `UPDATE public.headers SET check_count = (SELECT check_count WHERE id = $1) + 1 WHERE id = $1`
+	insertCheckedHeaderQuery = `
+INSERT INTO %s.checked_headers (check_count, header_id)
+VALUES (1, $1)
```
Comment (rmulhol) [inline pkg/datastore/postgres/repositories/checked_headers_repository.go:29]: I think we might want to retain the ability to mark the `check_count` as > 1, so that we can indicate rechecks

File: pkg/datastore/postgres/repositories/checked_headers_repository.go [hunk 3]
```diff
@@ -17,62 +17,77 @@
...
 	var (
 		result                  []core.Header
-		query                   string
 		err                     error
 		recheckOffsetMultiplier = 15
 	)
 
+	joinQuery := fmt.Sprintf(`SELECT h.id, h.block_number, h.hash
+						FROM public.headers h
+						LEFT JOIN %s.checked_headers ch
```
Comment (rmulhol) [inline pkg/datastore/postgres/repositories/checked_headers_repository.go]: Just making sure I follow - since we're doing a left join from the headers table, our result will include any headers where no row currently exists on the checked_headers table. Is that right?

File: pkg/datastore/postgres/repositories/checked_headers_repository.go [hunk 4]
```diff
@@ -17,62 +17,77 @@
...
-			FROM public.headers
-			WHERE (check_count < 1
-			           AND block_number >= $1)
-			   OR (check_count < $2
-			           AND block_number <= ((SELECT MAX(block_number) FROM public.headers) - ($3 * check_count * (check_count + 1) / 2)))`
-		err = repo.db.Select(&result, query, startingBlockNumber, checkCount, recheckOffsetMultiplier)
+		noEndingBlockQuery := fmt.Sprintf(`%s
+				 WHERE ((ch.check_count IS NULL OR ch.check_count < 1) AND h.block_number >= $1)
+				 OR ((ch.check_count IS NULL OR ch.check_count < $2)
+			           AND h.block_number <= ((SELECT MAX(block_number) FROM public.headers) - ($3 * ch.check_count * (ch.check_count + 1) / 2)))`, joinQuery)
```
Comment (rmulhol) [inline pkg/datastore/postgres/repositories/checked_headers_repository.go]: So seems like this may have been the case in the earlier implementation too, but does the second branch of this where clause disregard the starting block number?

File: test_config/test_config.go [hunk 5]
```diff
@@ -47,16 +45,6 @@ func setTestConfig() {
...
-	ipc := TestConfig.GetString("client.ipcPath")
-
-	// If we don't have an ipc path in the config file, check the env variable
-	if ipc == "" {
-		TestConfig.BindEnv("url", "CLIENT_IPCPATH")
-		ipc = TestConfig.GetString("url")
-	}
-	if ipc == "" {
-		logrus.Fatal(errors.New("testing.toml IPC path or $CLIENT_IPCPATH env variable need to be set"))
-	}
```
Comment (rmulhol) [inline test_config/test_config.go:59]: 😤

**Review by rmulhol — COMMENTED** [review]
Looking good!

File: pkg/datastore/postgres/repositories/checked_headers_repository.go (hunk 2 above)
Comment (paytonrules) [reply pkg/datastore/postgres/repositories/checked_headers_repository.go:29]: We did - that's how the ON_CONFLICT WORKS.  See line 32 where there is the +1.

What has changed is that entries in the join table aren't present until the first check, whereas before we relied on the header row having a default check_count of 0. Now we try to insert a check row, and if it's present (ON CONFLICT) we increment it by one.

This should be the same behavior as before, as the only place I saw where we explicitly set a value was in tests. Otherwise it was the same  "increment by 1" logic.

File: pkg/datastore/postgres/repositories/checked_headers_repository.go (hunk 3 above)
Comment (paytonrules) [reply pkg/datastore/postgres/repositories/checked_headers_repository.go]: Yes. We had good tests for this, fortunately, since I got it backwards the first time :).

File: pkg/datastore/postgres/repositories/checked_headers_repository.go (hunk 4 above)
Comment (paytonrules) [reply pkg/datastore/postgres/repositories/checked_headers_repository.go]: Looks like it does to me too. I hadn't noticed by I believe that's an error (albeit not a new one). Can try it out with a test.

File: pkg/datastore/postgres/repositories/checked_headers_repository.go (hunk 2 above)
Comment (rmulhol) [reply pkg/datastore/postgres/repositories/checked_headers_repository.go:29]: 😅 I feel very roasted by you putting "works" in all caps 🔥

File: integration_test/contract_watcher_transformer_test.go (hunk 1 above)
Comment (paytonrules) [reply integration_test/contract_watcher_transformer_test.go:38]: 🤦

File: pkg/datastore/postgres/repositories/checked_headers_repository.go (hunk 2 above)
Comment (paytonrules) [reply pkg/datastore/postgres/repositories/checked_headers_repository.go:29]: 😆 That was unintentional. I was just STILL TYPING THE SQL AND I GUESS KEPT GOING.

File: pkg/datastore/postgres/repositories/checked_headers_repository.go (hunk 4 above)
Comment (paytonrules) [reply pkg/datastore/postgres/repositories/checked_headers_repository.go]: I cleaned up this query a bit while fixing this bug, but I fixed it and added some tests.

File: cmd/execute.go [hunk 6]
```diff
@@ -96,7 +98,11 @@ func executeTransformers() {
 	// Use WaitGroup to wait on both goroutines
 	var wg sync.WaitGroup
 	if len(ethEventInitializers) > 0 {
-		extractor := logs.NewLogExtractor(&db, blockChain)
+		repo, repoErr := repositories.NewCheckedHeadersRepository(&db, genConfig.Schema)
+		if repoErr != nil {
+			LogWithCommand.Fatalf(fmt.Errorf("failed to create checked headers repository: %w", repoErr).Error())
```
Comment (rmulhol) [inline cmd/execute.go]: nbd but if we're just logging the error and not returning it, we could probably simplify this as:
```
LogWithCommand.Fatalf("failed to create checked headers repository: %s", repoErr.Error())
```

File: cmd/backfillEvents.go [hunk 7]
```diff
@@ -50,7 +51,11 @@ func backFillEvents() error {
 	blockChain := getBlockChain()
 	db := utils.LoadPostgres(databaseConfig, blockChain.Node())
 
-	extractor := logs.NewLogExtractor(&db, blockChain)
+	repo, repoErr := repositories.NewCheckedHeadersRepository(&db, genConfig.Schema)
+	if repoErr != nil {
+		return fmt.Errorf("error creating checked headers repository %w", repoErr)
```
Comment (rmulhol) [inline cmd/backfillEvents.go]: I _think_ this is probably a non-issue since we'll be running one container per schema, but maybe it'd be helpful to include the schema name in this formatted error?

File: libraries/shared/test_data/test_helpers.go [hunk 8]
```diff
@@ -88,3 +89,30 @@ func getRandomAddress() string {
 	stringHash := hashToPrefixedString(hash)
 	return stringHash[:42]
 }
+
+/// DescribeAValidCheckedHeadersModelWithSchema is provided so that
```
Comment (rmulhol) [inline libraries/shared/test_data/test_helpers.go]: seems like this comment is lying about the function name 🕵️

File: libraries/shared/test_data/test_helpers.go [hunk 9]
```diff
@@ -88,3 +89,30 @@ func getRandomAddress() string {
...
+	// insert header
+	blockNumber := rand.Int63()
+	fakeHeader := fakes.GetFakeHeader(blockNumber)
+	headerRepo := repositories.NewHeaderRepository(db)
+	headerID, headerErr := headerRepo.CreateOrUpdateHeader(fakeHeader)
+	Expect(headerErr).NotTo(HaveOccurred())
+
+	checkedHeaderRepo, repoErr := repositories.NewCheckedHeadersRepository(db, schema)
+	Expect(repoErr).NotTo(HaveOccurred())
+	uncheckedHeaders, uncheckedHeaderErr := checkedHeaderRepo.UncheckedHeaders(0, -1, 1)
```
Comment (rmulhol) [inline libraries/shared/test_data/test_helpers.go:118]: this is totally not your fault because it's how the function is implemented - but it might be cool to wrap `-1` (and potentially all of these numbers) in vars so that the intention is more expressive

File: pkg/datastore/postgres/repositories/checked_headers_repository.go [hunk 10]
```diff
@@ -17,62 +17,90 @@
...
-	db *postgres.DB
+	db         *postgres.DB
+	schemaName string
 }
 
-func NewCheckedHeadersRepository(db *postgres.DB) CheckedHeadersRepository {
-	return CheckedHeadersRepository{db: db}
+func NewCheckedHeadersRepository(db *postgres.DB, schemaName string) (*CheckedHeadersRepository, error) {
+	var sq SchemaQuery
+	err := db.Get(&sq, "SELECT schema_name FROM information_schema.schemata WHERE schema_name = $1", schemaName)
```
Comment (rmulhol) [inline pkg/datastore/postgres/repositories/checked_headers_repository.go:47]: 💪

File: pkg/datastore/postgres/repositories/checked_headers_repository_test.go [hunk 11]
```diff
@@ -30,313 +32,378 @@ import (
...
-			secondHeaderID,
-			thirdHeaderID int64
-			blockNumbers        []int64
-			headerIDs           []int64
-			err                 error
-			uncheckedCheckCount = int64(1)
-			recheckCheckCount   = int64(2)
-		)
+				Expect(repo.MarkHeaderChecked(headerID)).To(Succeed())
+				Expect(repo.MarkHeaderChecked(headerID)).To(Succeed())
```
Comment (rmulhol) [inline pkg/datastore/postgres/repositories/checked_headers_repository_test.go:85]: nbd but might we want a stronger assert than `Succeed()`? Seems like we could verify that the val is incremented

Comment (rmulhol) [reply pkg/datastore/postgres/repositories/checked_headers_repository_test.go:85]: lol lost in the diff, I see that the assert I'm describing is below 😅

**Review by rmulhol — APPROVED** [review]
Really awesome work here! None of my comments are merge blocking :shipit:

File: cmd/execute.go (hunk 6 above)
Comment (paytonrules) [reply cmd/execute.go]: 👍 - you have me trained to put fmt.Errorf everywhere!

File: cmd/backfillEvents.go (hunk 7 above)
Comment (paytonrules) [reply cmd/backfillEvents.go]: it should be in the error message, but it also shouldn't ever really happen so it won't hurt to stick in as well.

File: libraries/shared/test_data/test_helpers.go (hunk 8 above)
Comment (paytonrules) [reply libraries/shared/test_data/test_helpers.go]: Well at one time it was right. It's lying about how to use it too, so I'm fixing it.

File: libraries/shared/test_data/test_helpers.go (hunk 9 above)
Comment (paytonrules) [reply libraries/shared/test_data/test_helpers.go:118]: I might - although TBH I found some of these tests to be hard to follow because variables were in the setup and had to be understood at each level.

Maybe allowing a little WET in the tests would help with that.

File: pkg/datastore/postgres/repositories/checked_headers_repository_test.go (hunk 11 above)
Comment (paytonrules) [reply pkg/datastore/postgres/repositories/checked_headers_repository_test.go:85]: So I ended up changing a lot of this test because I learned to features of Gomega that I don't see a lot of in our tests.  It's documented here: 

https://onsi.github.io/gomega/#handling-errors

The short version is if you return value, err from a function (as you often do from Go funcs) and you use To(Equal(blah)) that ALSO validates that any other values in the tuple are nil. So you don't have to constantly check for error.

In addition above `Expect(repo.MarkHeaderChecked(headerID)).To(Succeed())` - requires all tuples to be nil returned from `MarkHeaderChecked`. This works best for functions like `MarkHeaderChecked` that only return an error.

I really like the first bit (we could remove a lot of explicit error checking, as it's redundant). The second I like for setup/action steps like MarkHeaderChecked is here, but I don't love it, and I'm really using it more like a traditional assert where something has gone horribly wrong if it fails.

Anyway I'm highlighting this because most of our tests aren't written this way, but I'll probably start migrating tests over to use less assertions when appropriate unless there's an objection.

Comment (rmulhol) [reply pkg/datastore/postgres/repositories/checked_headers_repository_test.go:85]: 👍 absolutely love it

**Review by elizabethengelman — APPROVED** [review]
👏
